### ISSUE TITLE
[Bug] Intentional Socket Closure during mod download

### ISSUE THREAD

**Issue by OfficialLambdax** [BeamMP/BeamMP-Server#125]
**Fill out general information**
OS (windows, linux, ...): Windows (maybe Linux)
BeamMP-**Launcher** Version: 2.0.77

**Describe the bug**
The launcher disconnects from the server during the download of mods from the server. This happens while the mod is downloaded not in between them. A wireshark capture and the server/client logs implie that the disconnect happens intentionally from the Launcher.

`[23/9/2022 00:01:57] [INFO] Launcher version is up to date
[23/9/2022 00:01:57] [INFO] Game Version : 0.26.1.0
[23/9/2022 00:01:57] [WARN] Game is newer than recommended, multiplayer may not work as intended!
[23/9/2022 00:01:57] [INFO] Downloading mod please wait...
[23/9/2022 00:01:58] [INFO] Download Complete!
[23/9/2022 00:01:58] [INFO] Game Launched!
[23/9/2022 00:02:13] [INFO] Game Connected!
[23/9/2022 00:02:13] [WARN] Game Reconnecting...
[23/9/2022 00:02:14] [INFO] Game Connected!
[23/9/2022 00:02:29] [INFO] Attempting to authenticate...
[23/9/2022 00:02:29] [INFO] Attempting to authenticate...
[23/9/2022 00:02:29] [INFO] Authentication successful!
[23/9/2022 00:02:30] [INFO] Authentication successful!
[23/9/2022 00:02:33] [INFO] Connecting to server
[23/9/2022 00:02:33] [INFO] Connected!
[23/9/2022 00:02:33] [INFO] Checking Resources...
[23/9/2022 00:02:33] [INFO] Syncing...
[23/9/2022 00:03:54] [INFO] 0                                                   <<<<<<<<------------------
[23/9/2022 00:03:54] [WARN] Failed to close socket!`

The Zero in the launcher log comes from:
https://github.com/BeamMP/BeamMP-Launcher/blob/master/src/Network/Resources.cpp#L132

which is the return of recv().
"If the connection has been **gracefully** closed, the return value is zero."
https://learn.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-recv

This means that the connection has been intentionally closed and the capture point towards the launcher causing it.

**To Reproduce**
I personally could not reproduce the bug, but these seem to be the steps to reproduce the bug. The bug does not happen each time. And seems to occure more often when the connecting player is far away from the server location (e.g. USA <-> GER)

1. Setup a Server
2. Put a bunch of Mods into the Client resource folder (it appears that the larger the mod, the more likely it is that the bug occurs e.g. a 700mb mod)
3. Connect to the Server with the game until the bug appears during mod download (you might have to delete the mods from your launcher resources folder to trigger the download again)

= You will see "Socket Closed Code 1" in the Lua mod and the 0 in the launcher when the bug has been reproduced successfully.

**Logs**
The Discord user Oro#9824 was so friendly to help debugging the bug so far and also uploaded the logs and capture here
https://www.dropbox.com/s/r3h7uom20dc91p5/Crash.7z?dl=0

**Additional context**
A ton of players already reported this bug in the dc by now

**Comment by Anonymous-275** [comment]
according to https://learn.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-recv
example code given reflects the fact that if the winsock recv function returns 0 the connection is therefore terminated

`iResult = recv(ConnectSocket, recvbuf, recvbuflen, 0);
        if ( iResult > 0 )
            printf("Bytes received: %d\n", iResult);
        else if ( iResult == 0 )
            printf("Connection closed\n");
        else
            printf("recv failed: %d\n", WSAGetLastError());`

**Comment by OfficialLambdax** [comment]
-1 is the return if the socket was closed unintentionally (connection lost for example), while the 0 clearly says that one of the two parties intented to close the socket

**Comment by Anonymous-275** [comment]
yes it is the only way to know, this check was added because of the fact that it would otherwise hang for an indefinite amount of time otherwise simply receiving 0, further strengthening the fact that 0 means closed.

**Comment by OfficialLambdax** [comment]
if it would get a -1 we would simply see the cause by calling WSAGetLastError

but since we get a 0 it points towards a bug, because that return only can be get when the socket has been closed intentionally. And in this case likely by the launcher, as the capture tells us
https://i.imgur.com/KFcMbd7.png

**Comment by Anonymous-275** [comment]
according to your picture the packet has the RST flag.

"In TCP, packets with the "Reset" (RST or R) flag are sent to abort a connection. Probably the most common reason you are seeing this is that an SYN packet is sent to a closed port. But RST packets may be sent in other cases to indicate that a connection should be closed."
- Google

**Comment by lionkor** [comment]
Closing a download socket gracefully should not stop anything, we should continue normally. No?

**Comment by Anonymous-275** [comment]
> Closing a download socket gracefully should not stop anything, we should continue normally. No?

we failed to get any data from the very socket the connection continues on, there is nothing left to salvage from that ;/

**Comment by Anonymous-275** [comment]
that is why we exploring new ideas on mod download, so that if the connection gets randomly cut we can try and resume for example :D

**Comment by OfficialLambdax** [comment]
Well currently it affects dozens of players that play on modded servers

**Comment by Anonymous-275** [comment]
thank you for the bug report, as I think it will be very helpful for us to mention it in the next dev meetings, I will see to it personally that the issue is fixed as soon as we decide as a team the new network designs.
